File: sections/42_advanced_data_structures_mixed/problems/h_connectivity_countdown/solution.py
import sys

def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    n, m, q = map(int, data[:3])
    index = 3
    edges = []
    for _ in range(m):
        u, v = map(int, data[index:index+2]); index += 2
        edges.append((u-1, v-1))
    operations = []
    deleted = [False] * m
    for _ in range(q):
        operation = data[index]; index += 1
        if operation == b'D':
            edge = int(data[index]) - 1; index += 1
            operations.append((operation, edge, -1)); deleted[edge] = True
        else:
            u, v = map(int, data[index:index+2]); index += 2
            operations.append((operation, u-1, v-1))
    parent = list(range(n)); size = [1] * n
    def find(u):
        while parent[u] != u:
            parent[u] = parent[parent[u]]; u = parent[u]
        return u
    def join(u, v):
        u, v = find(u), find(v)
        if u == v: return
        if size[u] < size[v]: u, v = v, u
        parent[v] = u; size[u] += size[v]
    for edge, (u, v) in enumerate(edges):
        if not deleted[edge]: join(u, v)
    answers = []
    for operation, x, y in reversed(operations):
        if operation == b'D': join(*edges[x])
        else: answers.append('YES' if find(x) == find(y) else 'NO')
    print('\n'.join(reversed(answers)))

if __name__ == "__main__":
    main()

File: sections/42_advanced_data_structures_mixed/problems/h_connectivity_countdown/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct DSU {
    vector<int> parent, size;
    DSU(int n) : parent(n), size(n, 1) { iota(parent.begin(), parent.end(), 0); }
    int find(int u) { return parent[u] == u ? u : parent[u] = find(parent[u]); }
    void join(int u, int v) {
        u = find(u); v = find(v);
        if (u == v) return;
        if (size[u] < size[v]) swap(u, v);
        parent[v] = u; size[u] += size[v];
    }
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);
    int n, m, q;
    if (!(cin >> n >> m >> q)) return 0;
    vector<pair<int, int>> edges(m);
    for (auto &[u, v] : edges) { cin >> u >> v; --u; --v; }
    struct Operation { char type; int x, y; };
    vector<Operation> operations(q);
    vector<char> deleted(m);
    for (auto &[type, x, y] : operations) {
        cin >> type >> x; --x;
        if (type == 'D') { y = -1; deleted[x] = true; }
        else { cin >> y; --y; }
    }
    DSU dsu(n);
    for (int e = 0; e < m; ++e)
        if (!deleted[e]) dsu.join(edges[e].first, edges[e].second);
    vector<string> answers;
    for (int i = q - 1; i >= 0; --i) {
        auto [type, x, y] = operations[i];
        if (type == 'D') dsu.join(edges[x].first, edges[x].second);
        else answers.push_back(dsu.find(x) == dsu.find(y) ? "YES" : "NO");
    }
    reverse(answers.begin(), answers.end());
    for (const string &answer : answers) cout << answer << '\n';
}
